src/service/data_products/filtering.py

```diff
@@ -0,0 +1,167 @@
+data products like genome attributes or samples.
+"""
+
+from dateutil import parser
+from pydantic import BaseModel, Field
+from src.common.product_models.columnar_attribs_common_models import ColumnType
+
+from types import NotImplementedType
+from typing import Any, Self
+
+
+class SearchQueryPart(BaseModel):
+    aql_lines: list[str] = Field(  # TDOO better docs once I figure out integration
+        description="One or more lines of ArangoSearch AQL representing the filter")
+    bind_vars: dict[str, Any] = Field(
+        description="The bind variables for the AQL lines")
+
+
+def _to_bool_string(b: bool):
+    return "true" if b else "false"
+
+
+class RangeFilter:
+    
+    def __init__(
+        self,
+        type_: ColumnType,
+        low: float | int | str | None = None,
+        high: float | int | str | None = None,
+        low_inclusive: bool = False,
+        high_inclusive: bool = False
+    ):
+        """
+        Create a range filter for numbers or ISO8601 dates.
+        
+        At least one of low or high must be supplied.
+        
+        type - the type of the range. Only ints, floats, and dates are allowed.
+        low - the low end of the range.
+        high - the high end of the range.
+        low_inclusive - whether to include the low value in the range.
+        high_inclusive - whether to include the high value in the range.
+        """
+        if type_ not in [ColumnType.DATE, ColumnType.INT, ColumnType.FLOAT]:
+            raise ValueError(f"Invalid type for range filter: {type_.value}")
+        self.type = type_
+        self.low = self._parse_val(low, "low")
+        self.high = self._parse_val(high, "high")
+        if self.low is None and self.high is None:
+            raise ValueError("At least one of the low or high values must be provided")
+        self.low_inclusive = low_inclusive
+        self.high_inclusive = high_inclusive
+        if self.low is not None and self.high is not None and (
+            self.low > self.high
+            or (self.low == self.high and (not self.low_inclusive or not self.high_inclusive))
+        ):
+            raise ValueError(f"The range {self.to_range_string()} excludes all values")
+
+    def to_range_string(self):
+        """
+        Create a range string from this instance. See `from_string` for a description of the
+        format.
+        """
+        ret = ""
+        if self.low is not None:
+            ret += "[" if self.low_inclusive else "("
+            ret += str(self.low)
+        ret += ","
+        if self.high is not None:
+            ret += str(self.high)
+            ret += "]" if self.high_inclusive else ")"
+        return ret
+
+    @classmethod
+    def _parse_inclusivity(self, part, start):
+        inclusive = False
+        if start:
+            if part.startswith("["):
+                inclusive = True
+            if part.startswith("(") or part.startswith("["):
+                part = part[1:]
```

Comment:
raise if range string is not None and doesn't start/end with `[` or `(`?

Reply:
The string specification says that `(` and `)` can be omitted without changing the semantics

Reply:
👍